Segment loading: Allow cancellation and prioritization of load queue items

kfaraz opened the pull request:
Fixes #12881

### Description

This PR lays the ground work to allow load queue to safely have an unlimited number of items, and thus
eventually phase out `maxSegmentsInNodeLoadingQueue` and `replicationThrottleLimit`.

Load queue is already allowed to have unlimited items (by setting `maxSegmentsInNodeLoadingQueue = 0`)
but this leads to
- each coordinator run taking a very long time
- poor assignments which take several more runs to be rectified

### Changes

#### Classes to review
- `StrategicSegmentAssigner`
- `LoadRule`, `BroadcastDistributionRule`, `DropRule`, `Rule.SegmentActionHandler`
- `SegmentLoadQueueManager`
- `LoadQueuePeon`: http and curator
- `SegmentHolder`
- `ServerHolder`

#### Behavioral changes

Change | Motivation
--------|-------------
Both loaded and loading items count towards replication. | Allow coordinator to take corrective action of removing superfluous replicas without waiting for them to be fully loaded.
Load, drop or move operations can be cancelled. | - Allow move of loading items from queue of one server to another. <br> - Allow coordinator to take corrective actions quickly.
During tier shift, always maintain the currently configured level of replication, no matter which tier it happens to be on. | - Queue drop of unneeded segments as soon as possible, thus allowing faster decommission of servers and freeing up disk space to load new segments.<br>- Always maintain target level of replication, thus ensuring that segment read concurrency does not suffer
LoadQueuePeon can distinguish and prioritize segment actions as DROP > LOAD > REPLICATE > MOVE (i.e. balancing). | - Allow prioritization of items, which becomes important if load queue size is unlimited. <br> - Avoid considering balancing items in load queue as over-replicated.
`replicationThrottleLimit` does not act on a tier if the segment is not loaded on that tier at all | Throttling first replica on a tier undermines the purpose of tiering. Tiering is not meant for fault tolerance, rather serving different query needs. Thus segments should be available on target tiers as soon as possible.
`maxNonPrimaryReplicantsToLoad` does not act on first replica in any tier | This was done keeping in line with the changes to `replicationThrottleLimit`
num items in load queue at start of run + num items assigned to load queue during run must be <=`maxSegmentsInNodeLoadingQueue` | Currently, if the configured load queue size is large enough to allow load of some segments while a coordinator run is in progress, the load queue limit is violated as there is always some room in the queue. This causes coordinator runs to get stuck cycling through all the segments in spite of a limited load queue.

#### Structural changes

Change | Motivation
--------|-----------
Add `StrategicSegmentAssigner` which handles all segment assignments. The lifecycle of the assigner is tied to a single coordinator run. | - Allow reuse of logic for loading, balancing and broadcasting. <br> - Single place to maintain state of a single run thus allowing better metrics and logging.
Load rules just specify their desired state and leave the actual decision making to the `StrategicSegmentAssigner`. | Simpler logic for load rules.
Add `SegmentLoadQueueManager` that interacts with the load queues. | - Single place to interact with load queue <br>- Allow reporting of metrics from queue callbacks. <br>- Prevent callbacks from holding references to items from the previous coordinator run.

#### New metrics
- `segment/loadQueue/assigned`
- `segment/loadQueue/success`
- `segment/loadQueue/cancelled`
- `segment/loadQueue/failed`
- `segment/assigned/broadcast`
- `segment/dropped/broadcast`
- Added dimension `"datasource"` in most segment level metrics

### Release notes

__--TODO--__

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/BalancerSegmentHolder.java:
```diff
@@ -19,29 +19,28 @@
 
 /**
  */
 public class BalancerSegmentHolder
```
Change summary:
Added `ServerHolder` to avoid having to find the corresponding peon later in the flow.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/BalancerStrategy.java:
```diff
@@ -46,13 +46,18 @@
    */
-  @Nullable
-  ServerHolder findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders);
+  Iterator<ServerHolder> findNewSegmentHomeReplicator(
```
Change summary:
Find all the best servers to place replicas in one go rather than multiple strategy computations.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java:
```diff
@@ -63,7 +56,7 @@
  */
 @Deprecated
-public class CuratorLoadQueuePeon extends LoadQueuePeon
+public class CuratorLoadQueuePeon implements LoadQueuePeon
```
Change summary:
Commoned out class `SegmentHolder` as `QueuedSegment` to represent an item in a load queue and to be used by both `HttpLoadQueuePeon` and `CuratorLoadQueuePeon`.
Implemented new methods in `LoadQueuePeon`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:
```diff
@@ -405,111 +405,6 @@ public String getCurrentLeader()
     return coordLeaderSelector.getCurrentLeader();
   }
 
-  public void moveSegment(
```
Change summary:
Moved this method to `SegmentLoader`/`SegmentStateManager`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/LoadQueuePeon.java:
```diff
@@ -21,37 +21,48 @@
  */
 @Deprecated
-public abstract class LoadQueuePeon
+public interface LoadQueuePeon
```
Change summary:
Added methods to `cancelLoad`, `cancelDrop`, and `loadSegment(segment, type)` (primary, replica or balance)

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/HttpLoadQueuePeon.java:
```diff
@@ -65,7 +66,7 @@
 /**
  */
-public class HttpLoadQueuePeon extends LoadQueuePeon
+public class HttpLoadQueuePeon implements LoadQueuePeon
```
Change summary:
Used `QueuedSegment` instead of `SegmentHolder` and implemented new `LoadQueuePeon` methods for cancellation and loading with type.

kfaraz replied:
Reverted.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CostBalancerStrategy.java:
```diff
@@ -184,20 +193,17 @@ public CostBalancerStrategy(ListeningExecutorService exec)
 
   @Override
-  public ServerHolder findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders)
+  public Iterator<ServerHolder> findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders)
```
Change summary:
Find all the best servers to place replicas in one go.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/ReplicationThrottler.java:
```diff
@@ -36,74 +39,89 @@
 {
   private static final EmittingLogger log = new EmittingLogger(ReplicationThrottler.class);
 
-  private final Map<String, Boolean> replicatingLookup = new HashMap<>();
```
Change summary:
Commoned out `ReplicationSegmentHolder` as `TierLoadingState` to be used for tracking both `currentlyReplicatingSegments` and `currentlyMovingSegments` (in `SegmentStateManager`).

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java:
```diff
@@ -20,31 +20,61 @@
 /**
+ *
  */
 public class ServerHolder implements Comparable<ServerHolder>
```
Change summary:
Maintain a map containing
- segments that were in load or drop queue when this run started
- segments that were queued for load or drop during this run
This helps the `SegmentLoader` determine the current state of a segment on a server and make load/drop decisions.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/BalanceSegments.java:
```diff
@@ -26,85 +26,65 @@
 
 /**
  */
 public class BalanceSegments implements CoordinatorDuty
```
Change summary:
Move state of `currentlyMovingSegments` to `SegmentStateManager`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/EmitClusterStatsAndMetrics.java:
```diff
@@ -234,36 +218,40 @@ private void emitStatsForHistoricalManagementDuties(DruidCluster cluster, Coordi
 
+    // Log load queue status of all replication or broadcast targets
     log.info("Load Queues:");
-    for (Iterable<ServerHolder> serverHolders : cluster.getSortedHistoricalsByTier()) {
```
Change summary: Report queue status not just of historicals but all replication or broadcast targets.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/EmitClusterStatsAndMetrics.java:
```diff
@@ -322,18 +310,14 @@ private void emitStatsForHistoricalManagementDuties(DruidCluster cluster, Coordi
-              )
-          );
-        }
+    coordinator.computeNumsUnavailableUsedSegmentsPerDataSource().forEach(
```
Style change: Use simpler iterator.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/RunRules.java:
```diff
@@ -46,38 +45,16 @@ public class RunRules implements CoordinatorDuty
 
-  private final DruidCoordinator coordinator;
-
-  public RunRules(DruidCoordinator coordinator)
```
Change summary:
Move replication throttling logic to `SegmentStateManager`.
Move check on `maxNonPrimaryReplicantsToLoad` to `ReplicationThrottler` (requires revisit).

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/simulate/SegmentLoadingNegativeTest.java:
```diff
@@ -91,130 +88,4 @@ public void testImmediateLoadingViolatesThrottleLimit()
-   * <p>
-   * Fix Apache #12881 to fix this test.
-   */
-  @Test
```
Change summary: 
The underlying behaviour for these tests is now fixed!
The corresponding positive test cases can be found in `SegmentLoadingTest`.

kfaraz reviewed (commented):
Added change summary to important classes.
The tests have been updated to use new flow but the same behaviour is still being verifiied.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentAction.java:
```diff
@@ -0,0 +1,33 @@
+public enum SegmentAction
+{
+  DROP,
+  LOAD_AS_PRIMARY,
```
Rename to something like `ACTIVE_LOAD`, `LAZY_LOAD`.

kfaraz replied:
Renamed to `LOAD`, `PRIORITY_LOAD`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,235 @@
+/**
+ * Manages state of segments being loaded.
+ */
+public class SegmentStateManager
```
Rename to `LoadQueueManager` as this class really acts as an interface between the duties and the load queues.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CostBalancerStrategy.java:
```diff
@@ -49,6 +50,14 @@ public class CostBalancerStrategy implements BalancerStrategy
+   */
+  private static final Comparator<Pair<Double, ServerHolder>> CHEAPEST_SERVERS_FIRST
+      = Comparator.<Pair<Double, ServerHolder>, Double>comparing(pair -> pair.lhs)
+      .thenComparing(pair -> ThreadLocalRandom.current().nextInt());
```
Sure, that would work too. 

This PR does not make any modifications to strategies, so it has not been included here. The only modification done here is reduction in the number of calls to strategy.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/server/coordinator/TierLoadingState.java:
```diff
@@ -0,0 +1,72 @@
+public class TierLoadingState
+{
+  private int lifetime;
+  private final ConcurrentHashMap<SegmentId, String> processingSegments = new ConcurrentHashMap<>();
```
Can't a tier load / balance the same segment on multiple servers simultaneously with these changes?

AmatyaAvadhanula replied:
The code doesn't enforce it, and it is possible for the same segment to be replicated / moved on different servers within the same tier.
This datastructure would lead to a loss of that information  (which also happens to be the case with the current implementation of ReplicationThrottler)

AmatyaAvadhanula replied:
A simple example is a LoadRule within a tier with a replication factor of 2. Both replicas can be loaded in the same cycle but the TierLoadingState would only keep track of the second load

kfaraz replied:
Yes, as you mention, that is the current implementation and we have retained it here.
We are hoping to relax replication throttle limit further anyway. So changing this behaviour is not really required.

Incidentally, it is also the desirable implementation, being simpler to understand and reason about than the alternative. From the docs, definition of `replicationThrottleLimit`:
`The maximum number of segments that can be replicated at one time.`
Alternative would be: 
`The number of replica segments (i.e. non primary) that can be in the queue at any time`.

We could try an alternative implementation but it would have these issues:
- Counting each replica separately would hit the throttle limit sooner (i.e. with fewer _unique_ segments), thus increasing the time to full replication. We could always increase the throttle limit but we would have to keep increasing it whenever we decided to increase the replication level on _any_ datasource.
- Assigning 1st replica of each segment, then 2nd replica of each segment is not an option either because we must make all decisions for a given segment in a single invocation of the `LoadRule.run()`.

Hope that clarifies the design choice 🙂

kfaraz replied:
However, this PR does modify the definition of `replicationThrottleLimit` and `maxSegmentsInNodeLoadingQueue` slightly (docs to be updated). The limits now apply on number of assignments done during a single coordinator run rather than items in the queue at a given instant.
This change was done to:
- make the configs easier to use
- ensure that the configs actually take effect. In some cases (where loading happens quickly enough), these limits are constantly violated.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/TierLoadingState.java:
```diff
@@ -0,0 +1,72 @@
+ */
+public class TierLoadingState
+{
+  private int lifetime;
```
That would be the ideal behaviour but as it turns out, this is the existing implementation. I have tried to keep it unchanged for now.

Let's say, in a coordinator run, we queue some segments to move within a tier. All the segments queued would start with the same lifecycle. Only when all these queued segments have been processed will we consider this tier for moving again. So, at it turns out, in `BalanceSegments.currentlyMovingSegments` in the existing code, all segment holders for a given tier will always have the same lifecycle.

AmatyaAvadhanula replied:
Please consider the usage in ReplicationThrottler where a TierLoadingState is added to `currentlyReplicating`.
A tier may be busy for a period greater than the lifetime without the load queue being stuck. In such a case the lifetime would still be decreased and may become < 0.

AmatyaAvadhanula replied:
Yes

kfaraz replied:
This is the current code:
https://github.com/apache/druid/blob/45dfd679e92e668172c470c37615e7447b601af1/server/src/main/java/org/apache/druid/server/coordinator/ReplicationThrottler.java#L73-L98

Here too a tier may be busy for a period greater than the lifetime, the lifetime would keep decreasing and we would keep alerting for this tier.

Saying that the load queue is stuck is probably not the correct message here and is causing the confusion. But AFAICT, the behaviour is unchanged.

Or are you talking about a `TierLoadingState` that doesn't have any segments inside and is empty? That has been handled like this:
```
// Reset state for inactive tiers
inactiveTiers.forEach(currentlyReplicating::remove);
```
The next time when a replica comes for this tier, a new entry would be created with a fresh lifetime value. The same handling has been done for moving segments as well.

AmatyaAvadhanula replied:
Apologies, the situation I'm thinking of can occur with the existing code as well.
I was thinking of timedOut in LoadQueue while making thing comment.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/HttpLoadQueuePeon.java:
```diff
@@ -470,134 +486,98 @@ public Set<DataSegment> getSegmentsMarkedToDrop()
-        return false;
-      }
-    }
+    return System.currentTimeMillis() - holder.getFirstRequestTimeMillis()
```
I fear that this logic is broken.  The call to `getFirstRequestTimeMillis()` ends up actually setting the current timestamp.  But, I don't know why this check is only actually done once the request is sent to the downstream server.  Rather than relying on such side-effects we would be better off having an explicit lifecycle to the assignment and asking the holder itself how long it has been queued at the server.  I.e. `holder.getTimeElapsedSinceFirstRequestToServer()` or something like that, which can return `null` if there hasn't been any request to the server (and thus should never be timed out)

kfaraz replied:
Yes, that would be cleaner.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/RandomBalancerStrategy.java:
```diff
@@ -26,24 +26,20 @@
 {
   @Override
-  public ServerHolder findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders)
+  public Iterator<ServerHolder> findNewSegmentHomeReplicator(DataSegment proposalSegment, List<ServerHolder> serverHolders)
```
Sometimes returning an `Iterator<>` is right, this signature feels like we might as well just return the `List<>` instead?

kfaraz replied:
Sounds good.

kfaraz replied:
On second thought, I wonder if `Iterator` is actually better. A caller should always iterate over the eligible servers in order and we get the benefit of the simple `hasNext` and `next` methods, rather than having to use `isEmpty` and `get`.

The one benefit I see with having a list is if we wanted to get the count of servers that were found to be eligible upfront (in case we wanted to report that).
What do you advise?

imply-cheddar replied:
I don't feel overly strong about it.  Was more of a stream-of-conciousness/feeling comment.  So now that you've thought about it directly and like Iterator better, let's keep that.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java:
```diff
@@ -0,0 +1,472 @@
+    // Find the minimum number of segments required for fault tolerance
+    final int totalTargetReplicas = tierToReplicaCount.values().stream()
+                                                      .reduce(0, Integer::sum);
+    final int minLoadedSegments = totalTargetReplicas > 1 ? 2 : 1;
+
+    // Drop segment from unneeded tiers if requirement is met across target tiers
+    int loadedTargetReplicas = 0;
+    final Set<String> targetTiers = tierToReplicaCount.keySet();
+    for (String tier : targetTiers) {
+      loadedTargetReplicas += replicantLookup.getLoadedReplicants(segment.getId(), tier);
+    }
+    if (loadedTargetReplicas < minLoadedSegments) {
+      return;
+    }
```
I'm perhaps missing something here, but the logic seems to be saying "if the replication factor is set higher than 1, make sure that there are 2 replicas and no more", which seems... weird.  We should have the target number that the user told us to have?

kfaraz replied:
This is the check for queueing up drops in unwanted tiers. If there is even the minimum level of fault tolerance (2 replicas) in the wanted tiers, we can proceed with the drops on unwanted tiers while full replication continues on the wanted tiers.

The assignment to new target tiers is already done by this point.

kfaraz replied:
Fixed this to always maintain the currently configured level of replication.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java:
```diff
@@ -0,0 +1,472 @@
+   */
+  public boolean canLoadSegment(ServerHolder server, DataSegment segment)
+  {
+    return server.canLoadSegment(segment)
```
I find myself wondering why the ServerHolder doesn't know about its own loadQueue?  Seems like this entire thing can be delegated.

kfaraz replied:
It does and I did start with that, in fact. But then did this because of laziness in fixing the tests to pass the config `maxSegmentsInNodeLoadingQueue` to every `ServerHolder`. 😅

I will fix this up.

kfaraz replied:
Fixed.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java:
```diff
@@ -0,0 +1,472 @@
+        && stateManager.loadSegment(segment, server, true)) {
+      return true;
+    } else {
+      log.makeAlert("Failed to broadcast segment for [%s]", segment.getDataSource())
```
Be careful with messages in this code.  "Failed to broadcast segment" is ambiguous about whether there was an issue with the server actually downloading the segment (definitely not the case given the code here) versus an issue with the coordinator believing that it is safe to assign the segment (much more likely).  We should be very explicit about what it is that is happening and, also, what we might expect the end user to do about it.

kfaraz replied:
Yes, I am going through all the log messages to make sure they are less ambiguous.

imply-cheddar reviewed (commented):
I'm not done, but need to do something else and don't want the comments to be stuck in draft, so submitting for now to make the comments visible.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:
```diff
@@ -255,6 +253,8 @@ public DruidCoordinator(
     this.coordLeaderSelector = coordLeaderSelector;
     this.objectMapper = objectMapper;
     this.compactSegments = initializeCompactSegmentsDuty();
+    this.segmentStateManager =
+        new SegmentStateManager(serverInventoryView, segmentsMetadataManager, taskMaster.isHttpLoading());
```
Agreed, I do have that in a separate patch. Will include it here.

kfaraz replied:
Added.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CuratorLoadQueuePeon.java:
```diff
@@ -181,10 +183,10 @@ public int getNumberOfSegmentsInQueue()
   {
-    SegmentHolder segmentHolder = new SegmentHolder(segment, Action.LOAD, Collections.singletonList(callback));
-    final SegmentHolder existingHolder = segmentsToLoad.putIfAbsent(segment, segmentHolder);
+    QueuedSegment segmentHolder = new QueuedSegment(segment, action, callback);
```
Yes, I will rename `QueuedSegment` back to `SegmentHolder` as it is adding unnecessarily to the diff clutter.

kfaraz replied:
Reverted to use `SegmentHolder`.

kfaraz replied:
> That said, I would've expected the queue itself to return a QueuedSegment rather than something creating one of those before it's actually queued...

Do you mean like a `computeIfAbsent` rather than `putIfAbsent`?

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCluster.java:
```diff
@@ -138,7 +138,7 @@ public Set<ServerHolder> getBrokers()
   }
 
-  public Iterable<String> getTierNames()
+  public Set<String> getTierNames()
```
`DruidCluster` is immutable, and I ended up using this method just once in the new code. So I will just stick to using the original `Iterable<String>`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/BalancerStrategy.java:
```diff
@@ -46,13 +46,18 @@
   /**
-   * Find the best server on which to place a {@link DataSegment} replica according to the balancing strategy
+   * Finds the best servers on which to place a replica of the {@code proposalSegment}
+   * according to the balancing strategy.
```
Okay, I will try to update the javadoc to reflect this distinction. But, as it happens, the same method is used for both "new/unavailable" segments as well "replicating/available" segments. (this would go away if we do a round-robin assignment in the first cycle for every segment) 
So I guess the method name should just be `findSegmentHomes`?

kfaraz replied:
Updated.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:
```diff
@@ -986,6 +887,7 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
 
       stopPeonsForDisappearedServers(currentServers);
 
+      segmentStateManager.prepareForRun(params);
```
Thanks for the suggestion! I was not happy with this myself.

kfaraz replied:
Fixed this to ensure that `SegmentStateManager` can never be used with improper initialization.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/duty/EmitClusterStatsAndMetrics.java:
```diff
@@ -43,41 +42,31 @@ public class EmitClusterStatsAndMetrics implements CoordinatorDuty
+      DruidCoordinator coordinator,
+      String groupName,
+      boolean isContainCompactSegmentDuty,
+      ServiceEmitter emitter
```
There is a single DI'd instance of `ServiceEmitter` that gets passed around. The emitter itself and the dimension map that it carries are immutable, so we should be good there.

Unfortunately, we didn't have any existing test that verified (all) the dimensions that were being emitted in a metric event. Such tests passing would have helped us validate this fully. But at least looking at the code, we shouldn't be breaking anything.

I will be doing some further cluster testing today to further verify the emitted metric events. We should be able to add some sim tests for this too.

AmatyaAvadhanula commented on server/src/main/java/org/apache/druid/server/coordinator/TierLoadingState.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Contains segments being loaded or replicated in a tier.
```
Isn't it segments being moved or being replicated, but not as a primary replica?

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/HttpLoadQueuePeon.java:
```diff
@@ -366,8 +372,9 @@ public void stop()
-  public void loadSegment(DataSegment segment, LoadPeonCallback callback)
+  public void loadSegment(DataSegment segment, SegmentAction action, LoadPeonCallback callback)
   {
+    Preconditions.checkArgument(action != SegmentAction.DROP);
```
Fixed.

kfaraz commented:
@imply-cheddar , thanks a lot for your review! I have addressed your comments.
More improvements to logging, metrics and docs will be included in a quick follow-up PR.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,261 @@
+  {
+    final String tier = server.getServer().getTier();
+    final LoadPeonCallback callback;
+    if (isPrimary) {
+      // Primary replicas are not subject to throttling
+      callback = null;
```
What's the definition of "isPrimary"?  There really isn't a notion of "primary" or "secondary" replication, at least there shouldn't be.  The only semantic difference that should exist is

1) If the data is not available at all inside of the tier -> get the data loaded as fast as humanly possibly
2) If the data is available somewhere inside of the tier -> throttle the loading so that there is capacity in the cluster for other activities as well as under the assumption that this usually happens due to a failure and the node could just be flapping.

kfaraz replied:
Yes, I have made this change in a follow up PR which fully enables prioritisation and does away with the notion of primary and secondary. Should I include that here itself?

kfaraz replied:
Fixed.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,261 @@
+    }
+
+    TierLoadingState tierState = currentlyReplicatingSegments.get(tier);
+    return tierState == null || tierState.getNumProcessingSegments() < throttler.getReplicationThrottleLimit();
```
The latter part of this really "feels" like it's a throttling decision, but the throttler appears to not be making that decision?  I don't know that I have a specific suggestion, just, it caught my eye as I read the code.

kfaraz replied:
Fair point. I will clean it up.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,261 @@
+      final TierLoadingState replicatingInTier = currentlyReplicatingSegments
+          .computeIfAbsent(tier, t -> new TierLoadingState(throttler.getMaxLifetime()));
+      replicatingInTier.addSegment(segment.getId(), server.getServer().getHost());
+      callback = success -> replicatingInTier.removeSegment(segment.getId());
```
Is this just internal bookkeeping?  Or is it actually asking for the segment to be removed?

If it's internal bookkeeping, `.markCompleted()` might be a better name.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,261 @@
+
+      server.getPeon().loadSegment(
+          segment,
+          isPrimary ? SegmentAction.LOAD_AS_PRIMARY : SegmentAction.LOAD_AS_REPLICA,
```
Let's please kill this false notion of primary/replica and use words that actually align with the semantics that should exist.

kfaraz replied:
Okay, I will include those changes in this PR itself. I am going with the term LOAD vs PRIORITY_LOAD.

kfaraz replied:
Fixed.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentStateManager.java:
```diff
@@ -0,0 +1,261 @@
+
+            // Do not check the inventory with http loading as the HTTP
+            // response is enough to determine load success or failure
+            if (success
```
If it wasn't successful, do we need to update the state that was set from the call to `fromServerPeon.markSegmentToDrop`?

kfaraz replied:
Yes, it happens through the `moveFinishCallback.execute()` in the `else` block.

imply-cheddar commented on server/src/test/java/org/apache/druid/server/coordinator/RunRulesTest.java:
```diff
@@ -119,35 +118,39 @@ public void testOneTierTwoReplicantsWithStrictReplicantLimit()
 
-    Assert.assertEquals(34L, stats.getTieredStat("assignedCount", "normal"));
-    Assert.assertEquals(10L, stats.getGlobalStat("totalNonPrimaryReplicantsLoaded"));
+    // There are 24 under-replicated segments, but only 10 replicas are assigned
```
Is there a stat that can be looked at and asserted on the 24 number?

kfaraz replied:
Added.

kfaraz commented on server/src/test/java/org/apache/druid/server/coordinator/BalancerStrategyTest.java:
```diff
@@ -84,29 +84,32 @@ public void findNewSegmentHomeReplicatorNotEnoughSpace()
+    Assert.assertFalse(
+        balancerStrategy.findNewSegmentHomeReplicator(proposedDataSegment, serverHolders)
+                        .hasNext()
+    );
```
Removed as we don't need the comment anymore.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicantLookup.java:
```diff
@@ -156,4 +161,45 @@ public Object2LongMap<String> getBroadcastUnderReplication(SegmentId segmentId)
+
+    void addQueued(SegmentState state)
+    {
+      switch (state) {
```
Fixed.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java:
```diff
@@ -0,0 +1,460 @@
+      }
+    }
+
+    // Drop as many replicas as possible from decommissioning servers
```
Thanks for the suggestion. Will include this changes along with the changes for full node replication in a follow up PR.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentLoader.java:
```diff
@@ -0,0 +1,460 @@
+    if (fromServer.isServingSegment(segment)) {
+      // Segment is loaded on fromServer, move it to toServer
+      return stateManager.moveSegment(segment, fromServer, toServer, replicationThrottler.getMaxLifetime());
+    } else if (!fromServer.isLoadingSegment(segment)) {
```
Yeah, it does make reading a little difficult. Fixed this here and a couple other places where I had used them.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentHolder.java:
```diff
@@ -0,0 +1,172 @@
+    return firstRequestMillis.get() > 0;
+  }
+
+  public long getMillisSinceFirstRequestToServer()
```
Thanks for the suggestion!

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentAction.java:
```diff
@@ -0,0 +1,33 @@
+  DROP,
+  LOAD,
+  REPLICATE,
+  MOVE_TO,
```
Added.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/HttpLoadQueuePeon.java:
```diff
@@ -380,10 +407,12 @@ public void loadSegment(DataSegment segment, LoadPeonCallback callback)
-        segmentsToLoad.put(segment, new LoadSegmentHolder(segment, callback));
+        queuedSize.addAndGet(segment.getSize());
+        holder = new SegmentHolder(segment, action, callback);
+        segmentsToLoad.put(segment, holder);
```
We might still need the `segmentsToLoad` and `segmentsToDrop` atleast for now because the balancer strategies use these to compute cost. This can be done with `queuedSegments` itself but we might have to filter out the relevant entries on every cost computation. 

I do have a follow up PR which deals with the fixes in the strategy. I will try to clean up this part there.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:
```diff
@@ -888,26 +799,101 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+      if (!dynamicConfig.isSmartSegmentLoading()) {
+        return dynamicConfig;
       }
 
-      return params.buildFromExisting()
-                   .withDruidCluster(cluster)
-                   .withLoadManagementPeons(loadManagementPeons)
-                   .withSegmentReplicantLookup(segmentReplicantLookup)
-                   .withRoundRobinServerSelector(roundRobinServerSelector)
-                   .build();
+      // Impose a lower bound on both replicationThrottleLimit and maxSegmentsToMove
+      final int throttlePercentage = 2;
+      final int replicationThrottleLimit = Math.max(
+          100,
+          params.getUsedSegments().size() * throttlePercentage / 100
+      );
+
+      // Impose an upper bound on maxSegmentsToMove to ensure that coordinator
+      // run times are bounded. This limit can be relaxed as performance of
+      // the CostBalancerStrategy.computeCost() is improved.
+      final int maxSegmentsToMove = Math.min(1000, replicationThrottleLimit);
+
+      log.info(
+          "Smart segment loading is enabled. Recomputed replicationThrottleLimit"
+          + " [%d] (%d%% of used segments) and maxSegmentsToMove [%d].",
+          replicationThrottleLimit, throttlePercentage, maxSegmentsToMove
+      );
+
+      return CoordinatorDynamicConfig.builder()
+                                     .withMaxSegmentsInNodeLoadingQueue(0)
+                                     .withReplicationThrottleLimit(replicationThrottleLimit)
+                                     .withMaxSegmentsToMove(maxSegmentsToMove)
+                                     .withUseRoundRobinSegmentAssignment(true)
+                                     .withUseBatchedSegmentSampler(true)
+                                     .withEmitBalancingStats(false)
```
Instead of recomputing the dynamic config, I would've preferred that there be a separate config which houses all of the things that are recomputed and that reference gets updated/changed.  The `DynamicConfig` object itself should really only be mutated by users changing config values.

kfaraz replied:
Yeah, I was not too keen on this myself. I considered putting this logic inside `DruidCoordinatorRuntimeParams` itself but I guess the best thing to do is just have a separate config object, as you suggest.

kfaraz replied:
Added a new `SegmentLoadingConfig` which contains the recomputed values and is used by the coordinator duties.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:
```diff
@@ -888,26 +799,101 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+     * be done before initializing the SegmentReplicantLookup so that
+     * under-replicated segments can be assigned in the current run itself.
+     */
+    private void cancelLoadsOnDecommissioningServers(DruidCluster cluster)
+    {
+      final AtomicInteger cancelledCount = new AtomicInteger(0);
+      final List<ServerHolder> decommissioningServers
+          = cluster.getAllServers().stream()
+                   .filter(ServerHolder::isDecommissioning)
+                   .collect(Collectors.toList());
+
+      for (ServerHolder server : decommissioningServers) {
+        server.getQueuedSegments().forEach(
+            (segment, action) -> {
+              // Cancel the operation if it is a type of load
+              if (action.isLoad() && server.cancelOperation(action, segment)) {
+                cancelledCount.incrementAndGet();
+              }
+            }
+        );
+      }
+
+      if (cancelledCount.get() > 0) {
+        log.info(
+            "Cancelled [%d] load/move operations on [%d] decommissioning servers.",
+            cancelledCount.get(), decommissioningServers.size()
+        );
+      }
     }
```
I would've expected that the act of decommissioning would put the queue into a state where it will only accept DROP requests and then dropped the load queue immediately.  It seems weird to me that this is logic handled in this class instead.  The only thing I'd expect a CoordinatorDuty to do about it is to look for decomissioning servers and move the segments away.

kfaraz replied:
Cancelling all loads on decomissioning servers is faster than moving the segments away.

Balancing moves are subject to the cost computation performance in `BalancerStrategy` and thus are typically limited (`maxSegmentsToMove <= 1000`). With unlimited load queues, there can potentially be many more segments in the load queue of decommissioning servers.

With cancellation of loads on decommissioning servers, these segments would immediately be (round-robin) assigned to active servers in this run itself.

kfaraz replied:
We are doing this cancellation in this duty (as opposed to `RunRules` or `BalanceSegments`) since the `SegmentReplicantLookup` is constructed right after this method and thus the Coordinator knows in this run itself that some segments are under-replicated and it needs to queue up some loads on active servers.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java:
```diff
@@ -155,128 +211,161 @@ public long getAvailableSize()
 
-    return server.getType().compareTo(serverHolder.server.getType());
+    return true;
+  }
```
nit: the boolean parameter here feels really weird.  Why not just have 2 methods:

```
addQueuedSegment()
removeQueuedSegment()
```

kfaraz replied:
Yeah, the boolean here is stupid, it's always `true`, must have been doing some other things originally. With the recent changes, it would have become redundant. Can get rid of it without breaking up the method too.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/StrategicSegmentAssigner.java:
```diff
@@ -0,0 +1,577 @@
+    return numLoadsQueued;
+  }
+
+  private boolean loadSegment(DataSegment segment, ServerHolder server, boolean isAlreadyLoadedOnTier)
+  {
+    final String tier = server.getServer().getTier();
+    if (isAlreadyLoadedOnTier && !replicationThrottler.canAssignReplica(tier)) {
+      incrementSkipStat(Stats.Segments.ASSIGN_SKIPPED, "Throttled replication", segment, tier);
+      return false;
+    }
+
+    final SegmentAction action = isAlreadyLoadedOnTier ? SegmentAction.REPLICATE : SegmentAction.LOAD;
+    final boolean assigned = loadQueueManager.loadSegment(segment, server, action);
+
+    if (!assigned) {
+      incrementSkipStat(Stats.Segments.ASSIGN_SKIPPED, "Encountered error", segment, tier);
+    } else if (isAlreadyLoadedOnTier) {
+      replicationThrottler.incrementAssignedReplicas(tier);
+    }
+
+    return assigned;
+  }
```
boolean police again: is it really better to have this boolean and a bunch of checks strewn through this code than to just have a `replicateSegment` method and a `loadSegment` method?

kfaraz replied:
The boolean here is used to determine if the assignment was successful or not, not distinguish between load and replicate. I can add a comment to that effect. The boolean status is used downstream for other decisions, like counting the number of successful assignments so far.

Please let me know if you think some other approach would be cleaner.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/StrategicSegmentAssigner.java:
```diff
@@ -0,0 +1,577 @@
+    return assigned;
+  }
+
+  private ReplicationThrottler createReplicationThrottler(CoordinatorDynamicConfig dynamicConfig)
```
Can this be static?  I think it can and maybe should be.  If it cannot be static, please rename to `initializeReplicationThrottler()` to help indicate that it is intended for initialization and not as a factory to be called multiple times from the same object.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/loadqueue/HttpLoadQueuePeon.java:
```diff
@@ -470,134 +517,127 @@
-      super.requestFailed(failureCause);
-    }
-  }
+      switch (action) {
```
This codeQL seems legitimate?

kfaraz replied:
No, it is okay, `MOVE_FROM` operations aren't supposed to come to the peon. So the next line will throw an IAE. I can just have the `default` case throw the exception too so that `CodeQL` does not complain.

imply-cheddar reviewed (approved):
Overall, I don't like the introduction of so many magical boolean parameters on methods.  It makes it more difficult to understand what is happening.  

Additionally, there is commentary in here, but I'm not sure any of them necessarily need to block this PR, it's all mostly just hygiene stuff that could be done later too (especially given that this PR is already so large and has existed for so long).

Given that we've run this in some legitimate production clusters as well as some performance environments and it's done what we expect, I'm approving this.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicantLookup.java:
```diff
@@ -125,49 +117,85 @@ public int getMovingReplicas(SegmentId segmentId, String tier)
+  {
+    ReplicaCount count = replicaCounts.get(segmentId, tier);
+    return (count == null) ? 0 : count.loadedNotDropping();
+  }
```
Yes, that would be neater, thanks!

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CoordinatorDynamicConfig.java:
```diff
@@ -428,66 +489,36 @@ public boolean equals(Object o)
-    }
-    if (!Objects.equals(dataSourcesToNotKillStalePendingSegmentsIn, that.dataSourcesToNotKillStalePendingSegmentsIn)) {
-      return false;
-    }
-    if (!Objects.equals(decommissioningNodes, that.decommissioningNodes)) {
-      return false;
-    }
-    if (pauseCoordination != that.pauseCoordination) {
-      return false;
-    }
-    if (replicateAfterLoadTimeout != that.replicateAfterLoadTimeout) {
-      return false;
-    }
-    if (maxNonPrimaryReplicantsToLoad != that.maxNonPrimaryReplicantsToLoad) {
-      return false;
-    }
-    return decommissioningMaxPercentOfMaxSegmentsToMove == that.decommissioningMaxPercentOfMaxSegmentsToMove;
+    return markSegmentAsUnusedDelayMillis == that.markSegmentAsUnusedDelayMillis
+           && mergeBytesLimit == that.mergeBytesLimit
+           && mergeSegmentsLimit == that.mergeSegmentsLimit
+           && maxSegmentsToMove == that.maxSegmentsToMove
+           && percentOfSegmentsToConsiderPerMove == that.percentOfSegmentsToConsiderPerMove
+           && decommissioningMaxPercentOfMaxSegmentsToMove == that.decommissioningMaxPercentOfMaxSegmentsToMove
+           && useBatchedSegmentSampler == that.useBatchedSegmentSampler
+           && balancerComputeThreads == that.balancerComputeThreads
+           && emitBalancingStats == that.emitBalancingStats
+           && replicantLifetime == that.replicantLifetime
+           && replicationThrottleLimit == that.replicationThrottleLimit
+           && replicateAfterLoadTimeout == that.replicateAfterLoadTimeout
+           && maxSegmentsInNodeLoadingQueue == that.maxSegmentsInNodeLoadingQueue
+           && maxNonPrimaryReplicantsToLoad == that.maxNonPrimaryReplicantsToLoad
+           && useRoundRobinSegmentAssignment == that.useRoundRobinSegmentAssignment
+           && pauseCoordination == that.pauseCoordination
+           && Objects.equals(
+               specificDataSourcesToKillUnusedSegmentsIn,
+               that.specificDataSourcesToKillUnusedSegmentsIn)
+           && Objects.equals(
+               dataSourcesToNotKillStalePendingSegmentsIn,
+               that.dataSourcesToNotKillStalePendingSegmentsIn)
+           && Objects.equals(decommissioningNodes, that.decommissioningNodes);
```
Thanks for catching this, must have missed adding it.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/CoordinatorDynamicConfig.java:
```diff
@@ -194,21 +211,46 @@ public CoordinatorDynamicConfig(
+      nameToDimension.put(dimension.reportedName(), dimension);
     }
+
+    debugDimensions.forEach(
+        (dimensionName, value) -> {
+          Dimension dimension = nameToDimension.get(dimensionName);
+          if (dimension != null && value != null) {
+            validDebugDimensions.put(dimension, value);
+          }
+        }
+    );
```
Don't know what I had done it in such a round about way in the first place 😅 , thanks for pointing this out!

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicantLookup.java:
```diff
@@ -125,49 +117,85 @@ public int getMovingReplicas(SegmentId segmentId, String tier)
-    return perTier;
+
+    return tierToUnderReplicated;
   }
```
Hmm, let me see what can be done here. 

It turned out like this mostly because it is a direct translation of the flag `computeUsingClusterView` passed in from `OverlordResource` to `DruidCoordinator` to `SegmentReplicantLookup`.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/balancer/BalancerStrategy.java:
```diff
@@ -36,77 +39,65 @@
       Set<String> broadcastDatasources,
-      int reservoirSize
+      int maxSegmentsToPick,
+      boolean pickLoadingSegments
```
Updated:
1. Removed these methods from `BalancerStrategy`
2. Added 2 methods to `ReservoirSegmentSampler`, `pickMovableLoadingSegment` and `pickMovableLoadedSegments`, which are used directly by the coordinator duty

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicantLookup.java:
```diff
@@ -125,49 +117,85 @@ public int getMovingReplicas(SegmentId segmentId, String tier)
+      totalLoaded += includeDropping ? count.loaded : count.loadedNotDropping();
     }
     return totalLoaded;
   }
```
Updated.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java:
```diff
@@ -337,29 +347,42 @@ private SegmentAction simplify(SegmentAction action)
+      // MOVE_FROM actions graduate to DROP after the corresponding MOVE_TO has finished
+      // Do not consider size delta until then, otherwise we might over-assign the server
     }
+  }
```
After the change to remove teh `addToQueue` boolean, you now have the exact same if clause at the beginning of 2 subsequence if statements.  You should be able to make it

```
if (action.isLoad()) {
  projectedSegments.add(segment);
  sizeOfLoadingSegments += segment.getSize();
} else {
  projectedSegments.remove(segment);
  // The current else block could be from a DROP or a MOVE_FROM.  The MOVE_FROM will eventually graduate to
  // DROP after ...
  if (action == SegmentAction.DROP) {
    sizeOfDroppingSegments += segment.getSize()
  }
}
```

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/ServerHolder.java:
```diff
@@ -337,29 +347,42 @@ private SegmentAction simplify(SegmentAction action)
+    } else if (action == SegmentAction.DROP) {
+      sizeOfDroppingSegments -= segment.getSize();
+    }
   }
```
Same comments here, basically have the same if statement twice.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/balancer/ReservoirSegmentSampler.java:
```diff
@@ -21,28 +21,86 @@
+   * @return Iterator over {@link BalancerSegmentHolder}s, each of which contains
+   * a segment picked for moving and the server currently loading it.
+   */
+  public static List<BalancerSegmentHolder> pickMovableLoadingSegmentsFrom(
```
nit: the different in name is soooo minute it took me a long time to realize.  One is `ing` and the other is `ed`.

Is there a reason not to expose the 4-argument method as `public` and have the call sites pass in one of the two lambdas?  It looks like there's only a single call-site for each of them except for tests.

That or maybe try to make the names a bit more different from each other.

imply-cheddar commented on server/src/main/java/org/apache/druid/server/coordinator/loading/SegmentLoadingConfig.java:
```diff
@@ -0,0 +1,124 @@
+  private final boolean useRoundRobinSegmentAssignment;
+  private final boolean emitBalancingStats;
+
+  public SegmentLoadingConfig(CoordinatorDynamicConfig dynamicConfig, int numUsedSegments)
```
design nit: a constructor should generally define the "dependency relationship" of a class.  I.e. the things passed in on the constructor are the things that the current class is dependent upon.  This constructor is doing a bunch of work, that work is dependent on the `CoordinatorDynamicConfig` object, but `SegmentLoadingConfig` is not dependent on the object.

In this case, it would be preferable for this to be a static method and the constructor just take all of the various values.

cheddar reviewed (approved):
making my gray check green.
